api/v1alpha1/virtualmachine_types.go

```diff
@@ -354,6 +354,39 @@ type VirtualMachineSpec struct {
 
 	// AdvancedOptions describes a set of optional, advanced options for configuring a VirtualMachine
 	AdvancedOptions *VirtualMachineAdvancedOptions `json:"advancedOptions,omitempty"`
+
+	// MinHardwareVersion specifies the desired, minimum hardware version
+	// for this VM.
+	//
+	// If omitted, the VM's hardware version is derived from the image used to
+	// deploy the VM. This field is never updated to reflect the derived
+	// hardware version. Instead, VirtualMachineStatus.HardwareVersion surfaces
+	// the observed hardware version.
+	//
+	// Please note, setting this field's value to N ensures a VM's hardware
+	// version is equal to or greater than N. For example, if a VM's observed
+	// hardware version is 10 and this field's value is 13, then the VM will be
+	// upgraded to hardware version 13. However, if the observed hardware
+	// version is 17 and this field's value is 13, no change will occur.
+	//
+	// Several features are hardware version dependent, for example:
+	//
+	// * Persistent Volume Claims        >= 13
+	// * Dynamic Direct Path I/O devices >= 17
+	//
+	// Please refer to https://kb.vmware.com/s/article/1003746 for a list of VM
+	// hardware versions.
+	//
+	// It is important to remember that a VM's hardware version may not be
+	// downgraded and upgrading a VM deployed from an image based on an older
+	// hardware version to a more recent one may result in unpredictable
+	// behavior. In other words, please be careful when choosing to upgrade a
+	// VM to a newer hardware version.
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=13
+	// +kubebuilder:validation:Maximum=20
+	MinHardwareVersion int32 `json:"minHardwareVersion,omitempty"`
```

Comment:
Since this is an `int32` and is an `optional` field, do check against the value `0`, to determine that no minimum hardware version has been specified?  (As opposed to making this a pointer and checking against `null`).

Reply:
Hmm, good question. When is the validation executed? I do not know. Brb 😃

Reply:
It appears fine, at least via kubectl. I:

Reply:
Updating `vm.yaml` to be the following:

```yaml
apiVersion: vmoperator.vmware.com/v1alpha1
kind: VirtualMachine
metadata:
  name: my-vm
spec:
  minHardwareVersion: 12
```

and reapplying it results in the expected error:

```bash
$ kubectl apply -f vm.yaml
The VirtualMachine "my-vm" is invalid: spec.minHardwareVersion: Invalid value: 12: spec.minHardwareVersion in body should be greater than or equal to 13
```

Reply:
Anyway, I think this is fine. The entire purpose of `omitempty` is to *not* send a value if the value is empty. You make it `nil` when you want a type's empty value to be something that can be sent, for example, like when you _want_ to send a value of `0`. Since we never want to do that here, using `0` as the empty value should be fine. Clients should not send it, and so validation rules won't apply.

@vincepri , do you happen to know if the above is accurate? If `0` is the empty value, client-go won't send an actual `0`, right?

Reply:
@aruneshpa, the `// +omitempty` indicates a field will not be serialized when it is the empty value. For integers this is `0`, so we're good.

Reply:
Got it.  Yeah, we should be good then.  
Another clarification is that even if this field is empty, because it is a primitive type, the controllers (when we read and de-serialize the object in its type) will still populate it as the default value of `0`, so unset doesn't really mean much here, right?

Reply:
Yes, but we know zero isn't a valid hardware version, so that's not an issue. The only time it matter is if 0 *is* a valid value for some numeric field.